### PR TITLE
Implement dataset spec

### PR DESCRIPTION
**Pull request by vhf**
Implements the current state of the dataset spec.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,5 +1,8 @@
...
+const rdf = require('@rdfjs/data-model')
+const Dataset = require('./Dataset')
 
-module.exports = function datasetFactory (quads = [], factory) {
-  return new Dataset(quads, factory)
+function dataset (quads) {
+  return new Dataset(quads)
 }
+
+module.exports = Object.assign({ dataset }, rdf)
```
Comment (thewilkybarkid) [inline index.js:8]: Could this change be extracted @vhf and a new release tagged even if the rest isn't? Refs https://gitter.im/rdfjs/public?at=5df8f727e47fb31eb7cadfc0.

Comment (vhf) [reply index.js:8]: @thewilkybarkid yep, let me see what I can do!

Comment (thewilkybarkid) [reply index.js:8]: 🎉 Would you accept TypeScript definitions here too? I'll open at DefinitelyTyped otherwise afterwards.

Comment (vhf) [reply index.js:8]: TypeScript definitions would be great! 😊

Comment (thewilkybarkid) [reply index.js:8]: 👍 

(I can extract the change if you like too.)

**Comment by bergos** [comment]
Current `Dataset` interface can't be implemented in an interoperable way, cause there are different opinions how some of the methods should work. That's why I will close this PR.

**Pull request closed without merging**